### DISCUSSION TITLE
Stepping Code - Minor Speedup

### DISCUSSION THREAD

**Discussion by IhorNehrutsa** [CAP1Sup/Intellistep#40] [General]
**1/32**
```
static const float sineTable[SINE_VAL_COUNT] = {
     0,    200,    401,    600,    798,    995,   1188,   1379, 
  1567,   1750,   1930,   2105,   2275,   2439,   2597,   2750,
  2895,   3034,   3165,   3289,   3404,   3512,   3611,   3701,
  3783,   3855,   3918,   3972,   4016,   4050,   4075,   4090,

  4095,   4090,   4075,   4050,   4016,   3972,   3918,   3855,
  3783,   3701,   3611,   3512,   3404,   3289,   3165,   3034,
  2895,   2750,   2597,   2439,   2275,   2105,   1930,   1750,
  1567,   1379,   1188,    995,    798,    600,    401,    200,

     0,   -200,   -401,   -600,   -798,   -995,  -1188,  -1379,
 -1567,  -1750,  -1930,  -2105,  -2275,  -2439,  -2597,  -2750,
 -2895,  -3034,  -3165,  -3289,  -3404,  -3512,  -3611,  -3701,
 -3783,  -3855,  -3918,  -3972,  -4016,  -4050,  -4075,  -4090,

 -4095,  -4090,  -4075,  -4050,  -4016,  -3972,  -3918,  -3855,
 -3783,  -3701,  -3611,  -3512,  -3404,  -3289,  -3165,  -3034,
 -2895,  -2750,  -2597,  -2439,  -2275,  -2105,  -1930,  -1750,
 -1567,  -1379,  -1188,   -995,   -798,   -600,   -401,   -200
};
```
Let's look at the stepping code from low to up. 

1) To navigate the sineTable, we don't need to know the angular position(float value) of the motor, only arrayIndex.
2) We can store current motor position as integer value (currentStep), not a float value (currentAngle).
3) uint32_t currentStep  can store 2^32/(2^13) == 2^19== 524288 revolutions == 188743680 degreeses
4) to calculate arrayIndex = currentStep  &  (SINE_VAL_COUNT-1)
5) to calculate currentAngle = currentStep * microstepAngle

So we can use integer calculations instead of float.

**Comment by CAP1Sup** [comment]
Sounds good. In reality, we don't really need to do hardly any multiplications. We can just add to the currentAngle like usual (needed for correction and PID), but we can also add to the currentStep, allowing the complete cutout of a good amount of equations. However, we still need to use a int32_t for the currentStep, as the steps and angle can sometimes go negative, depending on where the motor was powered on. I'll add this later today

**Reply by IhorNehrutsa** [reply]
Oh, yes, I meant int32_t

**Reply by CAP1Sup** [reply]
@IhorNehrutsa could you review this? I messed something up. https://github.com/CAP1Sup/Intellistep/commit/b94fe7136b32b5654fd9fdc639c84688ea25bff0

**Reply by IhorNehrutsa** [reply]
47kHz

**Reply by CAP1Sup** [reply]
So I was thinking of ways to speed up the stepping routine and I thought about the constants (defined in the config file). For example, we could force all of the decimal numbers to floats, reducing the processing required to compute them. What do you think @IhorNehrutsa ?

**Comment by IhorNehrutsa** [comment]
Floating point operations are usually slower than integers.
Now I think we need to focus on the core functionality of Intellistep (PID + SimpleFOC SimpleFOCStudio mode), but keep in mind the speed.
